OCPQE-18948 | ci: Create cluster autoscaler with ocm terraform provider

sunzhaohua2 opened the pull request:
@xueli181114 PTAL
```
$ ginkgo -focus OCP-69137 tests/e2e                                                                                                         
Ginkgo detected a version mismatch between the Ginkgo CLI and the version of Ginkgo imported by your packages:
  Ginkgo CLI Version:
    2.9.1
  Mismatched package versions found:
    2.13.2 used by e2e

  Ginkgo will continue to attempt to run but you may see errors (including flag
  parsing errors) and should either update your go.mod or your version of the
  Ginkgo CLI to match.

  To install the matching version of the CLI run
    go install github.com/onsi/ginkgo/v2/ginkgo
  from a path that contains a go.mod file.  Alternatively you can use
    go run github.com/onsi/ginkgo/v2/ginkgo
  from a path that contains a go.mod file to invoke the matching version of the
  Ginkgo CLI.

  If you are attempting to test multiple packages that each have a different
  version of the Ginkgo library with a single Ginkgo CLI that is currently
  unsupported.
  
time="2024-01-12T16:01:24+08:00" level=info msg="Running against env staging with gateway url https://api.stage.openshift.com"
time="2024-01-12T16:01:24+08:00" level=info msg="Loaded cluster profile configuration from profile rosa-sts-ad : map[additional_sg_number:4 admin_enabled:true autoscale:true byok:true byovpc:true ccs:true channel_group:candidate cloud_provider:aws compute_machine_type:m5.2xlarge etcd_encryption:true fips:true hypershift:false imdsv2:required labeling:true major_version:4.14 multi_az:true oidc_config:un-managed private:false private_link:false product_id:rosa proxy:true region:ap-northeast-1 sts:true tagging:true unified_acc_role_path:/unified/ version:latest worker_disk_size:200 zones:]"
Running Suite: e2e tests suite - /Users/sunzhaohua/code/terraform-provider-rhcs/tests/e2e
=========================================================================================
Random Seed: 1705046481

Will run 1 of 53 specs
SSSSSSSSSSSSS
------------------------------
P [PENDING]
TF Test Verfication/Post day 1 tests Author:xueli-Critical-OCP-69145 @OCP-69145 @xueli Apply to change security group will be forbidden
/Users/sunzhaohua/code/terraform-provider-rhcs/tests/e2e/verfication_post_day1_test.go:244
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSStime="2024-01-12T16:01:24+08:00" level=info msg="Loaded cluster profile configuration from profile rosa-sts-ad : map[additional_sg_number:4 admin_enabled:true autoscale:true byok:true byovpc:true ccs:true channel_group:candidate cloud_provider:aws compute_machine_type:m5.2xlarge etcd_encryption:true fips:true hypershift:false imdsv2:required labeling:true major_version:4.14 multi_az:true oidc_config:un-managed private:false private_link:false product_id:rosa proxy:true region:ap-northeast-1 sts:true tagging:true unified_acc_role_path:/unified/ version:latest worker_disk_size:200 zones:]"
time="2024-01-12T16:01:24+08:00" level=info msg="Running terraform init against the dir /Users/sunzhaohua/code/terraform-provider-rhcs/tests/tf-manifests/rhcs/cluster-autoscaler"
time="2024-01-12T16:01:31+08:00" level=info msg="Running terraform apply against the dir: /Users/sunzhaohua/code/terraform-provider-rhcs/tests/tf-manifests/rhcs/cluster-autoscaler with args [-var skip_nodes_with_local_storage=true -var max_pod_grace_period=10 -var balancing_ignored_labels=[\"l1\",\"l2\"] -var scale_down={\"delay_after_add\":\"3h\",\"delay_after_delete\":\"3h\",\"delay_after_failure\":\"3h\",\"enabled\":true,\"unneeded_time\":\"1h\",\"utilization_threshold\":\"0.5\"} -var cluster_id=28n855d5e30qjpp8hglv1j3jl9im4o77 -var balance_similar_node_groups=true -var log_verbosity=1 -var pod_priority_threshold=-10 -var ignore_daemonsets_utilization=true -var max_node_provision_time=1h -var resource_limits={\"cores\":{\"max\":1},\"max_nodes_total\":10,\"memory\":{\"max\":1}}]"
time="2024-01-12T16:01:37+08:00" level=info msg="Running terraform output against the dir: /Users/sunzhaohua/code/terraform-provider-rhcs/tests/tf-manifests/rhcs/cluster-autoscaler"
time="2024-01-12T16:01:37+08:00" level=info msg="Running terraform output against the dir: /Users/sunzhaohua/code/terraform-provider-rhcs/tests/tf-manifests/rhcs/cluster-autoscaler"
time="2024-01-12T16:01:38+08:00" level=info msg="Running terraform destroy against the dir: /Users/sunzhaohua/code/terraform-provider-rhcs/tests/tf-manifests/rhcs/cluster-autoscaler"
args:  /usr/local/bin/terraform destroy -auto-approve -no-color -var cluster_id=28n855d5e30qjpp8hglv1j3jl9im4o77 -var skip_nodes_with_local_storage=true -var max_pod_grace_period=10 -var pod_priority_threshold=-10 -var ignore_daemonsets_utilization=true -var balancing_ignored_labels=["l1","l2"] -var scale_down={"delay_after_add":"3h","delay_after_delete":"3h","delay_after_failure":"3h","enabled":true,"unneeded_time":"1h","utilization_threshold":"0.5"} -var balance_similar_node_groups=true -var log_verbosity=1 -var max_node_provision_time=1h -var resource_limits={"cores":{"max":1},"max_nodes_total":10,"memory":{"max":1}}
•SSSSSSSSSSSSSS

Ran 1 of 53 Specs in 23.317 seconds
SUCCESS! -- 1 Passed | 0 Failed | 1 Pending | 51 Skipped
PASS
```

xueli181114 commented on tests/tf-manifests/rhcs/cluster-autoscaler/main.tf:
```diff
@@ -0,0 +1,61 @@
+resource "rhcs_cluster_autoscaler" "cluster_autoscaler" {
+  cluster                       = var.cluster_id
+  balance_similar_node_groups   = true
+  skip_nodes_with_local_storage = true
```
@sunzhaohua2 You are hardcode all values without set from variables, then it's impossible to test via service for negative scenarios. Please update to like `skip_nodes_with_local_storage =var.skip_nodes_with_local_storage`

xueli181114 commented on tests/utils/exec/cluster-autoscaler.go:
```diff
@@ -0,0 +1,129 @@
+	return nil
+}
+
+func (ca *ClusterAutoscalerService) Apply(createArgs *ClusterAutoscalerArgs, recordtfargs bool, extraArgs ...string) (string, error) {
```
What's the differences between `MagicImport` and `Apply`? Both of them are using `terraform apply`

amalykhi commented on tests/utils/exec/cluster-autoscaler.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+func (ca *ClusterAutoscalerService) Init(manifestDirs ...string) error {
+	ca.ManifestDir = CON.AWSVPCDir
```
```suggestion
	ca.ManifestDir = CON.ClusterAutoscalerDir
```

amalykhi commented on tests/utils/exec/cluster-autoscaler.go:
```diff
@@ -0,0 +1,129 @@
+
+}
+
+func (ca *ClusterAutoscalerService) MagicImport(createArgs *ClusterAutoscalerArgs, extraArgs ...string) error {
```
What is this function for?

sunzhaohua2 commented:
/hold

sunzhaohua2 commented:
/unhold

xueli181114 commented:
/ok-to-test

xueli181114 commented on tests/tf-manifests/rhcs/cluster-autoscaler/variable.tf:
```diff
@@ -0,0 +1,98 @@
+variable "cluster_id" {
```
Will you please run `terraform fmt` in the manifestdir?

xueli181114 commented on tests/e2e/cluster_autoscaler_day2_test.go:
```diff
@@ -0,0 +1,98 @@
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	ci "github.com/terraform-redhat/terraform-provider-rhcs/tests/ci"
+	cms "github.com/terraform-redhat/terraform-provider-rhcs/tests/utils/cms"
```
The package name is `cms` no need to apply another alias

xueli181114 commented on tests/e2e/cluster_autoscaler_day2_test.go:
```diff
@@ -0,0 +1,98 @@
+			_, err := caService.Destroy()
+			Expect(err).ToNot(HaveOccurred())
+		})
+		Context("Author:zhsun-Medium-OCP-69137 @zhsun", func() {
```
This is a positive case, let's put the importance to `High`

xueli181114 commented on tests/e2e/cluster_autoscaler_day2_test.go:
```diff
@@ -0,0 +1,98 @@
+		Context("Author:zhsun-Medium-OCP-69137 @zhsun", func() {
+			It("Author:zhsun-Medium-OCP-69137 - Create cluster autoscaler with ocm terraform provider", ci.Day2, ci.Medium, ci.FeatureClusterautoscaler, func() {
+				By("Check no cluster autoscaler in cluster")
+				if profile.ClusterAutoscalerEnabled {
```
If this is skipped, do we have another case to delete the existing autoscaler and create again?

xueli181114 commented on tests/e2e/cluster_autoscaler_day2_test.go:
```diff
@@ -0,0 +1,98 @@
+				Expect(err).ToNot(HaveOccurred())
+				caResponseBody, err := cms.RetrieveClusterAutoscaler(ci.RHCSConnection, clusterID)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(caResponseBody.Body().LogVerbosity()).To(Equal(caOut.LogVerbosity))
```
Seems only checked the `Logverbosity`, how about other options checking?

xueli181114 commented on tests/utils/cms/cms.go:
```diff
@@ -255,6 +255,29 @@ func RetrieveClusterMachinePool(connection *client.Connection, clusterID string,
 	return resp.Body(), nil
 }
 
+func CreateClusterAutoscaler(connection *client.Connection, clusterID string, body *cmv1.ClusterAutoscaler) (*cmv1.AutoscalerPostResponse, error) {
```
Will other cases use the CreateClusterAutoscaler by SDK directly?

sunzhaohua2 replied:
I use it to recover the original clusterautoscaler

xueli181114 commented:
How about make the title more meaningful instead of set the case ID? Let's include the case title in title.

sunzhaohua2 commented:
Addressed all the comments, tested works well 
```
$ ginkgo -focus OCP-69137 tests/e2e                                                                                                          
...
time="2024-01-31T18:45:28+08:00" level=info msg="Loaded cluster profile configuration from profile rosa-sts-pl : map[admin_enabled:false autoscale:false byok:false byovpc:true ccs:true channel_group:candidate cloud_provider:aws compute_machine_type:m5.xlarge etcd_encryption:false fips:false hypershift:false imdsv2:optional kms_key_arn:false labeling:false major_version:4.14 multi_az:false oidc_config:managed private:true private_link:true product_id:rosa proxy:false region:us-east-1 sts:true tagging:false unified_acc_role_path: version:latest zones:]"
Running Suite: e2e tests suite - /Users/sunzhaohua/code/terraform-provider-rhcs/tests/e2e
=========================================================================================
Random Seed: 1706697925

Will run 1 of 57 specs
SSSSSSSSSSSS
------------------------------
P [PENDING]
TF Test Verfication/Post day 1 tests Author:xueli-Critical-OCP-69145 @OCP-69145 @xueli Apply to change security group will be forbidden
/Users/sunzhaohua/code/terraform-provider-rhcs/tests/e2e/verfication_post_day1_test.go:245
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSStime="2024-01-31T18:45:28+08:00" level=info msg="Running terraform init against the dir /Users/sunzhaohua/code/terraform-provider-rhcs/tests/tf-manifests/rhcs/cluster-autoscaler"
time="2024-01-31T18:45:31+08:00" level=info msg="Running terraform apply against the dir: /Users/sunzhaohua/code/terraform-provider-rhcs/tests/tf-manifests/rhcs/cluster-autoscaler with args [-var max_node_provision_time=1h -var scale_down={\"delay_after_add\":\"3h\",\"delay_after_delete\":\"3h\",\"delay_after_failure\":\"3h\",\"enabled\":true,\"unneeded_time\":\"1h\",\"utilization_threshold\":\"0.5\"} -var cluster_id=293105vhpj1uv57v90tg2t0ivl8p2chc -var skip_nodes_with_local_storage=true -var log_verbosity=1 -var pod_priority_threshold=-10 -var ignore_daemonsets_utilization=true -var balance_similar_node_groups=true -var max_pod_grace_period=10 -var balancing_ignored_labels=[\"l1\",\"l2\"] -var resource_limits={\"cores\":{\"max\":1},\"max_nodes_total\":10,\"memory\":{\"max\":1}}]"
time="2024-01-31T18:45:37+08:00" level=info msg="Running terraform output against the dir: /Users/sunzhaohua/code/terraform-provider-rhcs/tests/tf-manifests/rhcs/cluster-autoscaler"
time="2024-01-31T18:45:37+08:00" level=info msg="Running terraform output against the dir: /Users/sunzhaohua/code/terraform-provider-rhcs/tests/tf-manifests/rhcs/cluster-autoscaler"
•SSSS

Ran 1 of 57 Specs in 11.095 seconds
SUCCESS! -- 1 Passed | 0 Failed | 1 Pending | 55 Skipped
PASS

Ginkgo ran 1 suite in 13.540716493s
Test Suite Passed 
```

xueli181114 commented on tests/e2e/cluster_autoscaler_day2_test.go:
```diff
@@ -0,0 +1,136 @@
+				recreateAutoscaler, err := cms.PatchClusterAutoscaler(ci.RHCSConnection, clusterID, clusterAutoScalerBodyForRecreate)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(recreateAutoscaler.Status()).To(Equal(http.StatusOK))
+				//Here should be clusterAutoscalerAfter.Status() == http.StatusOK && clusterAutoscalerStatusBefore == http.StatusNotFound
```
Let's remove commented code

xueli181114 commented on tests/e2e/cluster_autoscaler_day2_test.go:
```diff
@@ -0,0 +1,136 @@
+				Expect(err).NotTo(HaveOccurred())
+				Expect(recreateAutoscaler.Status()).To(Equal(http.StatusOK))
+				//Here should be clusterAutoscalerAfter.Status() == http.StatusOK && clusterAutoscalerStatusBefore == http.StatusNotFound
+				//There is a bug for autoscaler status when autoscaler is not exist, clusterAutoscalerStatusBefore return 0 but should be 404
```
Can you please attach the bug id?

xueli181114 commented:
/lgtm

xueli181114 commented:
/approve

xueli181114 commented:
/lgtm

sunzhaohua2 commented:
/retest
